include/Dialects/VectorExt/VectorExtOps.td

```diff
@@ -59,4 +59,72 @@ def VectorExt_YieldOp : VectorExt_Op<"yield", [
   let assemblyFormat = "attr-dict ($operands^ `:` type($operands))?";
 }
 
+def VectorExt_WarpSingleLaneOp : VectorExt_Op<"warp_single_lane",
+      [DeclareOpInterfaceMethods<RegionBranchOpInterface>,
+       SingleBlockImplicitTerminator<"vector_ext::YieldOp">,
+       RecursiveSideEffects]> {
+  let summary = "vector.warp_single_lane operation";
+  let description = [{
+    Operation that encapsulate code that should be ran on a single lane within a
+    warp. Any code present in the region would only be executed on first lane
+    based on the laneid paramter. The op assumes `laneid` operand is an id to
+    identify the warp and it goes from 0 to warp_size.
+
+    // TODO: Add warp size as an attribute and support multiple dimensions lane
+    ids.
+
+    Operands are vector values distributed on all threads that may be used by
+    the single lane execution. The matching region argument is a vector of all
+    the values of those threads available to the single active lane. The
+    distributed dimension is implicit based on the shape of the operand and
+    argument. In the future this may be described by an affine map.
+
+    Return values are distributed on all threads using laneId as index. The
+    vector is distributed based on the shape ratio between the vector type of
+    the yield and the result type.
+    If the shapes are the same this means the value is broadcasted to all lanes.
+    In the future the distribution can be made more explicit using affine_maps
+    and will support having multiple Ids.
+
+    During lowering values passed as operands and return value need to be
+    visible to different lanes within the warp. This would usually be done by
+    going through memory.
+
+    Example:
+    ```
+    vector_ext.warp_single_lane (%laneid) {
+      ...
+    }
+
+    %0 = vector_ext.warp_single_lane(%laneid)
+    args(%v0 : vector<4xi32>) -> (vector<1xf32>) {
```

Comment:
I am a little confused now as the stuff we had discussed offline seemed to:
1. capture from above (i.e. not be isolated)
2. take all other data from buffers

What you propose here makes sense too but seems to be a different design (which makes sense as things are evolving).
I would formalize the thinking a little more in the doc and explain how the distributed vectors and the sequential vectors relate. A slightly larger example connecting to function boundaries would help too.

I suggest something like:

```
// future: attribute or extra operand to describe #lanes
func @distributed_running_on_all_lanes(%laneid: index, %v: vector<4xi32>) -> vector<1xi32>{
  // Execute in parallel on all lanes.
  ...
  some_synchronization_primitive(...)
  // Alternatively, allow: vector_ext.warp_single_lane [%lane_id = 0] ... 
  %0 = vector_ext.warp_single_lane when %lane_id is %c0 args(%v0 : vector<4xi32>) -> (vector<1xf32>) {

      // 32 * vector<4xi32> are packed into a vector<128xi32> following the natural order 
      // This is the simplest form that makes sense.
      // This may have to go through memory (i.e. explicit buffers) when lowering to executable IR.
      // TODO: more advanced mechanisms than simple natural order packing. 
      ^bb0(%arg0 : vector<128xi32>) :
        // Execute sequentially only on lane 0.
        ...

        // vector<32xi32> unpack into 32 * vector<1xi32> following the natural order 
        // This is the simplest form that makes sense.
        // This may have to go through memory (i.e. explicit buffers) when lowering to executable IR.
        // TODO: more advanced mechanisms than simple natural order unpacking. 
        vector_ext.yield %1 : vector<32xf32>
    }

    some_synchronization_primitive(...)

    // Execute in parallel on all lanes.
}
```

The more advanced mechanisms for stacking / decomposing start to feel a bit like the indexing_maps of a linalg.generic on vectors but with some variations that we need to discover.
The packing/unpacking ops seem close to the vector.insert/extract_map which at this point could be renamed into distributed_packing / distributed_unpacking ?

Does the intuition of the above description make sense ?